### DISCUSSION TITLE
Installing certificates so that browser doesn't warn me every-time?

### DISCUSSION THREAD

**Discussion by flikhamud45** [synchronizing/mitm#8] [Q&A] [answered]
Does this package support certificates? And if so, how can I use this to install a certificate so the browser won't warn every time the software is running. 
Thank you very much.

**Comment by synchronizing** [comment]
Yup. The [`MITM`](https://synchronizing.github.io/mitm/module/mitm.html#mitm.mitm.MITM) class has an optional `ssl_context` variable that allows you to pass a custom [ssl.SSLContext](https://docs.python.org/3/library/ssl.html#ssl.SSLContext) which you can use to load your own custom certificates. Currently, `mitm` generates a new certificate on every run and it stores them under the following path (you need to run this via Python to figure it out - in my case):

```python
>>> import mitm
>>> mitm.__data__
PosixPath('/Users/felipe/Library/Application Support/mitm')
```

The certificates that are used are generated via [`mitm_ssl_default_context()`](https://github.com/synchronizing/mitm/blob/master/mitm/crypto.py#L118), and you can use that function to:

1. Generate a private/public key and store them on your desktop.
2. Install the certificates on your machine.
3. Load the certificates into `mitm` before run.

A similar question to this was answered in #3, but it related to an older version of `mitm` (which had the `Config()` class).

**Reply by flikhamud45** [reply]
the problem is that it uses the same certificate for every website so chrome gives an error that says that the certificate is not suitable for this website since every website needs a unique certificate. I tried to use my middleware object so I can react every time there is a connection to a new website but I can't change the certificate the MITM uses while running. Is there any way I can do so?

**Reply by synchronizing** [reply]
> the problem is that it uses the same certificate for every website so chrome gives an error that says that the certificate is not suitable for this website since every website needs a unique certificate

I'm honestly not sure how Chrome works in this regards, but I'll open an issue to attempt to make this step a little bit easier.

> I tried to use my middleware object so I can react every time there is a connection to a new website but I can't change the certificate the MITM uses while running.

Indeed, this wouldn't be possible in the current iteration of `mitm`. I'll open an issue in regards to this and do more research on specifically Chrome + Mac. Hopefully (and if you are willing) we can figure out the steps for Windows thereafter.

**Reply by synchronizing** [reply]
#11 opened. I'll likely only be able to work on this come this weekend or next week, unfortunately, but I'll keep you posted.

**Comment by synchronizing** [comment, marked as answer]
mitm version 1.3.0 has been officially released with support for certificate authority and dummy cert generation. See [here](https://github.com/synchronizing/mitm/issues/11#issuecomment-1133972481). 

Marking this as 'answered'. 😄
